images/ci-prune-buildcache/ci_buildcache_prune.py

```diff
@@ -0,0 +1,497 @@
+import argparse
+import boto3
+from collections import defaultdict
+from contextlib import contextmanager
+from datetime import datetime, timedelta, timezone
+import gitlab
+import multiprocessing as mp
+import multiprocessing.pool as pool
+import os
+from io import BytesIO
+import sentry_sdk
+from urllib.parse import urlparse
+
+import buildcache
+import helper
+from pruner import DirectPruner, IndexPruner, OrphanPruner
+
+sentry_sdk.init(
+    # This cron job only runs once weekly,
+    # so just record all transactions.
+    traces_sample_rate=1.0,
+)
+
+# Script Config
+PRUNE_REF = os.environ.get("PRUNE_REF", "develop")
+SINCE_DAYS = int(os.environ.get("PRUNE_SINCE_DAYS", 14))
+
+# Gitlab API (for fetching artifacts)
+GITLAB_URL = os.environ.get("GITLAB_URL", "https://gitlab.spack.io")
+GITLAB_PROJECT = os.environ.get("GITLAB_PROJECT", "spack/spack")
+gl = gitlab.Gitlab(GITLAB_URL)
+
+# S3
+BUILDCACHE_URL = os.environ.get("BUILDCACHE_URL", f"s3://spack-binaries/{PRUNE_REF}").rstrip("/")
+
+TODAY = datetime.today()
+TODAY_STR = TODAY.strftime("%Y.%m.%d")
+
+enable_delete = os.environ.get("AWS_ACCESS_KEY_ID", True)
+
+
+class Cache:
+    def __init__(self):
+        self.cache_dir = None
+
+    def write(self, key: str, stream):
+        if not self.cache_dir:
+            return
+
+        key_path = os.path.join(self.cache_dir, key)
+        if os.path.exists(key_path):
+            raise Exception(f"Duplicate cache key: {key_path}")
+
+        # This is happening in parallel, try it but don't fail
+        if not os.path.isdir(os.path.dirname(key_path)):
+            try:
+                os.makedirs(os.path.dirname(key_path))
+            except:
```

Comment:
```suggestion
            except Exception:
```

Reply:
I left it open since I don't care if/how it fails here. afaict the only ways this would fail is if a) I can't write to the fs which means the entire operation is borked anyway or b) the directory already exists in which case we are fine to continue.